File: scripts/import_roe.py
#!/usr/bin/env python3
"""
import_roe.py
=============
Merge ROE data into all_stocks_daily chunk files and stocks.db.

ROE source: /root/qlib_data/roe/{TICKER}.{EXCHANGE}.csv
  - Quarterly frequency, keyed by announcement date (date column)
  - Fields: date, end_date, symbol, roe, roe_ttm, roe_deducted
  - symbol format: sz000001 (lowercase)

Target:
  1. Chunk files: /projects/portal/data/quant/processed/all_stocks_daily_chunk{0,1,2}.csv
     -> Add roe_ttm column (forward-filled from announcement date)
  2. SQLite: /projects/portal/data/quant/processed/stocks.db  kline table
     -> Add roe_ttm column

Strategy: use announcement date (date) as the point-in-time marker,
forward-fill roe_ttm to each trading day to avoid look-ahead bias.

Usage:
    python scripts/import_roe.py
"""
from __future__ import annotations
import os, sys, glob, sqlite3, time
import pandas as pd
import numpy as np
from pathlib import Path

ROE_DIR   = "/root/qlib_data/roe"
CHUNK_DIR = "/projects/portal/data/quant/processed"
DB_PATH   = os.path.join(CHUNK_DIR, "stocks.db")
CHUNK_PATTERN = os.path.join(CHUNK_DIR, "all_stocks_daily_chunk*.csv")


def load_all_roe() -> pd.DataFrame:
    """Load all ROE CSVs and build a unified DataFrame."""
    files = sorted(glob.glob(os.path.join(ROE_DIR, "*.csv")))
    print(f"[roe] Found {len(files)} ROE files")

    frames = []
    for f in files:
        try:
            df = pd.read_csv(f, dtype=str)
            if df.empty or "roe_ttm" not in df.columns:
                continue
            frames.append(df[["date", "symbol", "roe_ttm"]])
        except Exception as e:
            print(f"  [warn] Skipping {f}: {e}")
            continue

    roe = pd.concat(frames, ignore_index=True)
    roe["date"] = pd.to_datetime(roe["date"])
    roe["roe_ttm"] = pd.to_numeric(roe["roe_ttm"], errors="coerce")
    # symbol is lowercase like "sz000001"
    # chunk files use lowercase, SQLite uses uppercase
    # Keep lowercase as canonical, convert when needed
    roe["code"] = roe["symbol"].str.lower()
    roe = roe.drop(columns=["symbol"])
    roe = roe.dropna(subset=["roe_ttm"])
    roe = roe.sort_values(["code", "date"]).drop_duplicates(subset=["code", "date"], keep="last")
    print(f"[roe] Loaded {len(roe):,} valid ROE observations for {roe['code'].nunique()} stocks")
    return roe


def forward_fill_roe_for_chunk(chunk_df: pd.DataFrame, roe_df: pd.DataFrame) -> pd.DataFrame:
    """
    For each stock in chunk_df, forward-fill roe_ttm from announcement dates.
    This is a point-in-time merge: on any trading day, the roe_ttm value is
    the most recent ROE that was publicly announced on or before that date.
    """
    chunk_df = chunk_df.copy()
    chunk_df["_date"] = pd.to_datetime(chunk_df["date"])

    # Build a lookup: for each code, sorted list of (announce_date, roe_ttm)
    roe_lookup = {}
    for code, grp in roe_df.groupby("code"):
        roe_lookup[code] = grp[["date", "roe_ttm"]].sort_values("date").values

    # Vectorized approach: merge_asof per code group
    # But with 22M rows and 5K stocks, we use merge_asof which is efficient
    chunk_df = chunk_df.sort_values(["code", "_date"])
    roe_df_sorted = roe_df.sort_values(["code", "date"])

    # merge_asof needs both sorted by the key
    results = []
    codes = chunk_df["code"].unique()
    total = len(codes)
    t0 = time.time()

    for i, code in enumerate(codes):
        stock_daily = chunk_df[chunk_df["code"] == code][["_date"]].copy()
        if code in roe_lookup:
            roe_vals = roe_lookup[code]
            # For each daily date, find the latest ROE announcement <= that date
            roe_series = pd.DataFrame(roe_vals, columns=["ann_date", "roe_ttm"])
            roe_series["ann_date"] = pd.to_datetime(roe_series["ann_date"])
            merged = pd.merge_asof(
                stock_daily.sort_values("_date"),
                roe_series.sort_values("ann_date"),
                left_on="_date",
                right_on="ann_date",
                direction="backward"
            )
            results.append(merged.set_index(stock_daily.index)["roe_ttm"])
        else:
            results.append(pd.Series(np.nan, index=stock_daily.index, name="roe_ttm"))

        if (i + 1) % 500 == 0 or i + 1 == total:
            elapsed = time.time() - t0
            print(f"\r  [ffill] {i+1}/{total} stocks ({elapsed:.0f}s)", end="", flush=True)

    print()
    chunk_df["roe_ttm"] = pd.concat(results)
    chunk_df = chunk_df.drop(columns=["_date"])
    return chunk_df


def process_chunks(roe_df: pd.DataFrame):
    """Add roe_ttm column to each chunk file."""
    chunk_files = sorted(glob.glob(CHUNK_PATTERN))
    print(f"\n[chunks] Found {len(chunk_files)} chunk files")

    for cfile in chunk_files:
        print(f"\n[chunk] Processing {os.path.basename(cfile)} ...")
        t0 = time.time()
        chunk = pd.read_csv(cfile, dtype=str)
        nrows = len(chunk)
        print(f"  Loaded {nrows:,} rows")

        # Remove existing roe_ttm column if present (idempotent)
        if "roe_ttm" in chunk.columns:
            chunk = chunk.drop(columns=["roe_ttm"])
            print("  Dropped existing roe_ttm column")

        # Ensure lowercase for matching (chunk files already lowercase)
        chunk["code"] = chunk["code"].str.lower()

        # Forward-fill ROE
        chunk = forward_fill_roe_for_chunk(chunk, roe_df)

        # Save
        chunk.to_csv(cfile, index=False)
        elapsed = time.time() - t0
        filled = chunk["roe_ttm"].notna().sum() if "roe_ttm" in chunk.columns else 0
        fill_pct = filled / nrows * 100 if nrows > 0 else 0
        print(f"  Saved. ROE coverage: {filled:,}/{nrows:,} ({fill_pct:.1f}%) in {elapsed:.0f}s")


def update_sqlite(roe_df: pd.DataFrame):
    """Add roe_ttm column to SQLite kline table."""
    print(f"\n[sqlite] Updating {DB_PATH} ...")
    conn = sqlite3.connect(DB_PATH)
    cur = conn.cursor()

    # Check if roe_ttm column already exists
    cols = [row[1] for row in cur.execute("PRAGMA table_info(kline)").fetchall()]
    if "roe_ttm" not in cols:
        print("  Adding roe_ttm column ...")
        cur.execute("ALTER TABLE kline ADD COLUMN roe_ttm REAL")
        conn.commit()
    else:
        print("  roe_ttm column already exists, will update values")

    # Strategy: for each stock, get all trading dates from kline,
    # then forward-fill ROE and batch update.
    # This is more efficient than row-by-row.

    # SQLite stores codes in UPPERCASE
    roe_upper = roe_df.copy()
    roe_upper["code"] = roe_upper["code"].str.upper()
    codes = roe_upper["code"].unique().tolist()
    print(f"  Updating ROE for {len(codes)} stocks ...")

    # Prepare batch update statement
    update_stmt = "UPDATE kline SET roe_ttm = ? WHERE code = ? AND date = ?"

    t0 = time.time()
    total_updated = 0

    for i, code in enumerate(codes):
        # Get trading dates for this stock
        rows = cur.execute(
            "SELECT date FROM kline WHERE code = ? ORDER BY date", (code,)
        ).fetchall()
        if not rows:
            continue

        dates = pd.DataFrame(rows, columns=["date"])
        dates["_date"] = pd.to_datetime(dates["date"])

        # Get ROE announcements for this stock (uppercase)
        stock_roe = roe_upper[roe_upper["code"] == code][["date", "roe_ttm"]].copy()
        stock_roe = stock_roe.rename(columns={"date": "ann_date"})
        stock_roe["ann_date"] = pd.to_datetime(stock_roe["ann_date"])
        stock_roe = stock_roe.sort_values("ann_date")

        if stock_roe.empty:
            continue

        # merge_asof
        merged = pd.merge_asof(
            dates.sort_values("_date"),
            stock_roe,
            left_on="_date",
            right_on="ann_date",
            direction="backward"
        )

        # Only update rows with non-null roe_ttm
        valid = merged.dropna(subset=["roe_ttm"])
        if valid.empty:
            continue

        updates = [(float(row["roe_ttm"]), code, row["date"]) for _, row in valid.iterrows()]

        cur.executemany(update_stmt, updates)
        total_updated += len(updates)

        if (i + 1) % 500 == 0 or (i + 1) == len(codes):
            conn.commit()
            elapsed = time.time() - t0
            print(f"\r  [sqlite] {i+1}/{len(codes)} stocks, {total_updated:,} rows updated ({elapsed:.0f}s)", end="", flush=True)

    conn.commit()
    print(f"\n  Total rows updated: {total_updated:,}")

    # Verify
    sample = cur.execute(
        "SELECT code, date, roe_ttm FROM kline WHERE roe_ttm IS NOT NULL ORDER BY date DESC LIMIT 5"
    ).fetchall()
    total_with_roe = cur.execute("SELECT COUNT(*) FROM kline WHERE roe_ttm IS NOT NULL").fetchone()[0]
    total_rows = cur.execute("SELECT COUNT(*) FROM kline").fetchone()[0]
    pct = total_with_roe / total_rows * 100 if total_rows > 0 else 0

    print(f"  Verification: {total_with_roe:,}/{total_rows:,} rows have ROE ({pct:.1f}%)")
    print(f"  Sample rows:")
    for row in sample:
        print(f"    {row[0]} | {row[1]} | roe_ttm={row[2]}")

    conn.close()
    print("[sqlite] Done.")


if __name__ == "__main__":
    print("=" * 64)
    print("ROE Data Import Script")
    print("=" * 64)

    print("\n[1/3] Loading all ROE data ...")
    roe = load_all_roe()

    print("\n[2/3] Merging into chunk files ...")
    process_chunks(roe)

    print("\n[3/3] Updating SQLite database ...")
    update_sqlite(roe)

    print("\n✅ All done!")

File: scripts/import_roe.mjs
#!/usr/bin/env node
/**
 * import_roe.mjs
 * ==============
 * Merge ROE data into:
 *   1. all_stocks_daily_chunk{0,1,2}.csv  (add roe_ttm column)
 *   2. stocks.db kline table              (add roe_ttm column)
 *
 * ROE source: /root/qlib_data/roe/{TICKER}.{EXCHANGE}.csv
 *   - Quarterly, keyed by announcement date
 *   - Fields: date, end_date, symbol, roe, roe_ttm, roe_deducted
 *   - symbol format: sz000001 (lowercase)
 *
 * Strategy: use announcement date as point-in-time marker,
 * forward-fill roe_ttm to each trading day (no look-ahead bias).
 *
 * Usage:
 *   node scripts/import_roe.mjs
 */

import Database from 'better-sqlite3';
import fs from 'fs';
import path from 'path';
import readline from 'readline';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = path.resolve(__dirname, '..');

const ROE_DIR = 'data/quant/processed/roe';
const CHUNK_DIR = path.join(PROJECT_ROOT, 'data/quant/processed');
const DB_PATH = path.join(CHUNK_DIR, 'stocks.db');

// ─── Step 1: Load all ROE data into memory ───────────────────────────────────

/**
 * Load all ROE CSVs and return a Map<code_upper, Array<{annDate, roeTtm}>>
 * sorted by annDate ascending.
 */
async function loadAllROE() {
  const files = fs.readdirSync(ROE_DIR).filter(f => f.endsWith('.csv')).sort();
  console.log(`[roe] Found ${files.length} ROE files`);

  // Map: uppercase code -> [{annDate: 'YYYY-MM-DD', roeTtm: number}]
  const roeMap = new Map();
  let totalObs = 0;

  for (const file of files) {
    const fpath = path.join(ROE_DIR, file);
    const lines = fs.readFileSync(fpath, 'utf8').split('\n');
    if (lines.length < 2) continue;

    // Parse header
    const header = lines[0].trim().split(',');
    const idx = {};
    header.forEach((col, i) => { idx[col.trim()] = i; });
    if (!('roe_ttm' in idx) || !('date' in idx) || !('symbol' in idx)) continue;

    for (let i = 1; i < lines.length; i++) {
      const line = lines[i].trim();
      if (!line) continue;
      const cols = line.split(',');
      const annDate = cols[idx['date']]?.trim();
      const symbol = cols[idx['symbol']]?.trim();
      const roeTtmStr = cols[idx['roe_ttm']]?.trim();

      if (!annDate || !symbol || !roeTtmStr || roeTtmStr === '') continue;
      const roeTtm = parseFloat(roeTtmStr);
      if (isNaN(roeTtm)) continue;

      const codeUpper = symbol.toUpperCase();
      if (!roeMap.has(codeUpper)) roeMap.set(codeUpper, []);
      roeMap.get(codeUpper).push({ annDate, roeTtm });
      totalObs++;
    }
  }

  // Sort each stock's ROE by announcement date
  for (const [code, arr] of roeMap) {
    arr.sort((a, b) => a.annDate.localeCompare(b.annDate));
  }

  console.log(`[roe] Loaded ${totalObs.toLocaleString()} valid observations for ${roeMap.size} stocks`);
  return roeMap;
}

/**
 * Given a sorted array of {annDate, roeTtm} and a target date string,
 * find the most recent ROE announced on or before that date.
 * Uses binary search for efficiency.
 */
function lookupROE(roeArr, dateStr) {
  if (!roeArr || roeArr.length === 0) return null;

  let lo = 0, hi = roeArr.length - 1;
  let result = null;

  while (lo <= hi) {
    const mid = (lo + hi) >>> 1;
    if (roeArr[mid].annDate <= dateStr) {
      result = roeArr[mid].roeTtm;
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }
  return result;
}

// ─── Step 2: Process chunk CSV files ─────────────────────────────────────────

async function processChunks(roeMap) {
  const chunkFiles = fs.readdirSync(CHUNK_DIR)
    .filter(f => f.startsWith('all_stocks_daily_chunk') && f.endsWith('.csv'))
    .sort()
    .map(f => path.join(CHUNK_DIR, f));

  console.log(`\n[chunks] Found ${chunkFiles.length} chunk files`);

  for (const chunkFile of chunkFiles) {
    console.log(`\n[chunk] Processing ${path.basename(chunkFile)} ...`);
    const t0 = Date.now();

    // Stream read → tmp file → rename (handles multi-GB files)
    const tmpFile = chunkFile + '.tmp';
    const writeStream = fs.createWriteStream(tmpFile, { encoding: 'utf8' });

    let headerParsed = false;
    let codeIdx = -1, dateIdx = -1, roeTtmIdx = -1;
    let filled = 0, total = 0;

    await new Promise((resolve, reject) => {
      const rl = readline.createInterface({
        input: fs.createReadStream(chunkFile, { encoding: 'utf8' }),
        crlfDelay: Infinity,
      });

      rl.on('line', (line) => {
        if (!line.trim()) return;

        if (!headerParsed) {
          headerParsed = true;
          const headers = line.trim().split(',');
          codeIdx = headers.indexOf('code');
          dateIdx = headers.indexOf('date');
          roeTtmIdx = headers.indexOf('roe_ttm');

          if (codeIdx === -1 || dateIdx === -1) {
            console.log('  Missing code/date columns, skipping');
            rl.close();
            return;
          }

          if (roeTtmIdx !== -1) {
            console.log('  roe_ttm column exists, will overwrite');
            writeStream.write(line + '\n');
          } else {
            console.log('  Adding roe_ttm column');
            writeStream.write(line + ',roe_ttm\n');
          }
          return;
        }

        const cols = line.split(',');
        const code = cols[codeIdx]?.trim();
        const date = cols[dateIdx]?.trim();

        if (!code || !date) {
          writeStream.write(line + (roeTtmIdx === -1 ? ',' : '') + '\n');
          return;
        }

        total++;
        const codeUpper = code.toUpperCase();
        const roeArr = roeMap.get(codeUpper);
        const roeTtm = lookupROE(roeArr, date);

        if (roeTtmIdx !== -1) {
          cols[roeTtmIdx] = roeTtm != null ? roeTtm.toString() : '';
          writeStream.write(cols.join(',') + '\n');
        } else {
          writeStream.write(line + ',' + (roeTtm != null ? roeTtm.toString() : '') + '\n');
        }

        if (roeTtm != null) filled++;

        if (total % 2_000_000 === 0) {
          const elapsed = ((Date.now() - t0) / 1000).toFixed(0);
          process.stdout.write(`\r  Processed ${total.toLocaleString()} rows (${elapsed}s)...`);
        }
      });

      rl.on('close', () => {
        writeStream.end(() => resolve());
      });
      rl.on('error', reject);
    });

    // Replace original with tmp
    fs.renameSync(tmpFile, chunkFile);

    const elapsed = ((Date.now() - t0) / 1000).toFixed(1);
    const pct = total > 0 ? (filled / total * 100).toFixed(1) : '0.0';
    console.log(`\n  Saved. ROE coverage: ${filled.toLocaleString()}/${total.toLocaleString()} (${pct}%) in ${elapsed}s`);
  }
}

// ─── Step 3: Update SQLite ───────────────────────────────────────────────────

function updateSQLite(roeMap) {
  console.log(`\n[sqlite] Updating ${DB_PATH} ...`);
  const db = new Database(DB_PATH);
  db.pragma('journal_mode = WAL');
  db.pragma('synchronous = NORMAL');
  db.pragma('cache_size = -65536');

  // Check if roe_ttm column exists
  const cols = db.pragma('table_info(kline)').map(c => c.name);
  if (!cols.includes('roe_ttm')) {
    console.log('  Adding roe_ttm column ...');
    db.exec('ALTER TABLE kline ADD COLUMN roe_ttm REAL');
  } else {
    console.log('  roe_ttm column already exists, will update values');
  }

  const selectDates = db.prepare('SELECT date FROM kline WHERE code = ? ORDER BY date');
  const updateStmt = db.prepare('UPDATE kline SET roe_ttm = ? WHERE code = ? AND date = ?');

  const codes = Array.from(roeMap.keys()); // uppercase
  console.log(`  Updating ROE for ${codes.length} stocks ...`);

  const t0 = Date.now();
  let totalUpdated = 0;

  // Batch in transaction
  const BATCH = 500;
  for (let batchStart = 0; batchStart < codes.length; batchStart += BATCH) {
    const batchEnd = Math.min(batchStart + BATCH, codes.length);
    const batchCodes = codes.slice(batchStart, batchEnd);

    const txn = db.transaction(() => {
      for (const code of batchCodes) {
        const roeArr = roeMap.get(code);
        if (!roeArr || roeArr.length === 0) continue;

        const dates = selectDates.all(code);
        for (const row of dates) {
          const roeTtm = lookupROE(roeArr, row.date);
          if (roeTtm != null) {
            updateStmt.run(roeTtm, code, row.date);
            totalUpdated++;
          }
        }
      }
    });
    txn();

    const elapsed = ((Date.now() - t0) / 1000).toFixed(0);
    process.stdout.write(`\r  [sqlite] ${batchEnd}/${codes.length} stocks, ${totalUpdated.toLocaleString()} rows updated (${elapsed}s)`);
  }

  console.log('');

  // Verify
  const totalWithROE = db.prepare('SELECT COUNT(*) as cnt FROM kline WHERE roe_ttm IS NOT NULL').get().cnt;
  const totalRows = db.prepare('SELECT COUNT(*) as cnt FROM kline').get().cnt;
  const pct = totalRows > 0 ? (totalWithROE / totalRows * 100).toFixed(1) : '0.0';
  console.log(`  Verification: ${totalWithROE.toLocaleString()}/${totalRows.toLocaleString()} rows have ROE (${pct}%)`);

  const samples = db.prepare(
    'SELECT code, date, roe_ttm FROM kline WHERE roe_ttm IS NOT NULL ORDER BY date DESC LIMIT 5'
  ).all();
  console.log('  Sample rows:');
  for (const s of samples) {
    console.log(`    ${s.code} | ${s.date} | roe_ttm=${s.roe_ttm}`);
  }

  db.exec('ANALYZE;');
  db.pragma('wal_checkpoint(TRUNCATE)');
  db.close();
  console.log('[sqlite] Done.');
}

// ─── Main ────────────────────────────────────────────────────────────────────

async function main() {
  console.log('='.repeat(64));
  console.log('ROE Data Import Script (Node.js)');
  console.log('='.repeat(64));

  console.log('\n[1/3] Loading all ROE data ...');
  const roeMap = await loadAllROE();

  console.log('\n[2/3] Merging into chunk files ...');
  await processChunks(roeMap);

  console.log('\n[3/3] Updating SQLite database ...');
  updateSQLite(roeMap);

  console.log('\n✅ All done!');
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});
